CORE-18088 Race condition registering RPC endpoints in JavalinServer

mbrkic-r3 opened the pull request:
Ensuring that JavalinServer is thread-safe by changing the`Set` implementation for registered endpoints in `JavalinServer` to prevent threading issues when multiple clients register endpoint at the same time as well as synchronising Javalin server starts across the process in case of multiple servers.

mbrkic-r3 committed:
CORE-18088 Using thread safe Set implementation for registered endpoints in JavalinServer

driessamyn reviewed (dismissed):
LGTM

mbrkic-r3 committed:
CORE-18088 Fixed unit tests

driessamyn commented:
I'm ok with this change is it's pretty harmless, but I'm 99% sure this will not fix the issue we saw.
This is because this log message:

```
2023-10-27T21:47:44.742261644Z stdout F {"instant":{"epochSecond":1698443264,"nanoOfSecond":741000000},"thread":"lifecycle-coordinator-0","level":"INFO","loggerName":"net.corda.web.server.JavalinServer","message":"Handler added to webserver: HandlerMetaInfo(httpMethod=POST, path=/api/5.1/ledger, handler=net.corda.web.server.JavalinServer$$Lambda$928/0x0000000801aaa000@3b3f335a, roles=[])","endOfBatch":false,"loggerFqcn":"org.apache.logging.slf4j.Log4jLogger","threadId":23,"threadPriority":5}
```

is generated by the Javalin registration handler, so won't be impacted by the thread safety of the Set that is maintained in `JavalinServer`

Commit:
Add locking around starting Javalin server so it only starts one at a time.
Also add locking around the registration of endpoints for the RPC endpoints. This should prevent Combined worker issues with multiple processors registering endpoints at the same time.

LWogan reviewed (dismissed):
LGTM

vkolomeyko reviewed (dismissed):
LGTM

LWogan commented on libs/web/web-impl/src/main/kotlin/net/corda/web/server/JavalinServer.kt:
```diff
@@ -58,48 +63,76 @@ class JavalinServer(
-                // The logging implementation is found correctly in practice.
-                executeWithStdErrSuppressed {
-                    server?.start(port)
+//                JavalinStarter.startServer(
```
we should remove this commented out code

LWogan commented on libs/web/web-impl/src/main/kotlin/net/corda/web/server/JavalinServer.kt:
```diff
@@ -58,48 +63,76 @@ class JavalinServer(
+                val bundle = FrameworkUtil.getBundle(WebSocketServletFactory::class.java)
+                val name = "HTTP RPC"
+                if (bundle != null) {
+//                    val bundleList = threadLocalClassLoaderBundles.plus(bundle)
```
remove

driessamyn replied:
I'm just trying something out to drill down into the issue, please ignore the last few commits.

LWogan commented on libs/web/web-impl/src/main/kotlin/net/corda/web/server/JavalinServer.kt:
```diff
@@ -68,38 +66,32 @@ class JavalinServer(
+        // Ensure the server can only be started once at a time.
+        serverStartLock.withLock {
+            log.info("Starting Worker Web Server on port: $port")
+            server = javalinFactory()
```
The startServer method is protected by this lock, but I think the stopServer() could be executed concurrently if two threads call restartServer(). This could mean that `server ` gets set to null after this point. There is a checkNotNull() as part of `registerEndpointInternal(it)` which would throw an IllegalStateException and bubble up unexpectedly, and would probably cause the worker to move to an ERROR state and be bounced by k8s.

This scenario could happen on the persistence worker which has 2 endpoints/SyncRPCSubscriptions. Its a bit of an edge case. If a dependent service causes the parent lifecycle coordinators, LedgerPersistenceService & FlowPersistenceService to receive a RegistrationStatusChangeEvent of DOWN, then they will both restart the http subscriptions and will both call restartServer().

As an aside (and I think its already a known issue), the "green path" of this use case would cause the JavalinServer to bounce 4 times. Twice for closing the 2 subs, and twice more for starting new ones.

driessamyn replied:
Good point.

I move the lock to the public start & stop methods and also the restart method.
We just need to be careful this doesn't create a potential deadlock scenario, but I think this should be ok.

Commit:
Move lock to public start/stop methods + restart.

LWogan reviewed (approved):
LGTM
